### PR TITLE
feat(parachain/availabilitystore): implement database store/retrieve

### PR DESCRIPTION
**Pull request by edwardmack**
## Changes
Implement functionality to store and retrieve Availability Data, Candidate Meta Data, and Erasure Chunk Data.

## Tests

```sh
go test github.com/ChainSafe/gossamer/dot/parachain/availability-store
```

## Issues
related to #3415 

## Primary Reviewer

@kishansagathiya 
@timwu20

### CODE REVIEW

File: dot/parachain/availability-store/messages.go
Comment (kishansagathiya) [inline dot/parachain/availability-store/messages.go:1]: let's add comments for all struct definitions, that will help us understand the code better when we revisit it in future

File: dot/parachain/availability-store/availabilitystore.go [hunk 1]
```diff
@@ -57,33 +172,92 @@ func (av *AvailabilityStoreSubsystem) processMessages() {
...
+		logger.Errorf("failed to load meta data: %w", err)
+	}
+	msg.Sender <- meta.ChunksStored[msg.ValidatorIndex]
 }
 
 func (av *AvailabilityStoreSubsystem) handleStoreChunk(msg StoreChunk) {
-	// TODO: handle store chunk
+	err := av.availabilityStore.StoreChunk(msg.CandidateHash, msg.Chunk)
+	if err != nil {
+		msg.Sender <- err
```
Comment (kishansagathiya) [inline dot/parachain/availability-store/availabilitystore.go:252]:
```suggestion
		msg.Sender <- err
		return
```

File: dot/node_integration_test.go [hunk 2]
```diff
@@ -120,7 +120,11 @@ func TestNewNode(t *testing.T) {
 		return stateSrvc, nil
 	})
 
-	phs, err := parachain.NewService(testNetworkService, "random_fork_id", common.Hash{})
+	mockBlockState := &state.BlockState{}
+	mockStateService := &state.Service{
+		Block: mockBlockState,
+	}
+	phs, err := parachain.NewService(testNetworkService, "random_fork_id", mockStateService)
```
Comment (kanishkatn) [inline dot/node_integration_test.go:127]: Just curious, what does `phs` stand for?

File: dot/parachain/availability-store/availabilitystore.go [hunk 3]
```diff
@@ -57,33 +172,92 @@ func (av *AvailabilityStoreSubsystem) processMessages() {
...
+	}
+	msg.Sender <- nil
 }
 
 func (av *AvailabilityStoreSubsystem) handleStoreAvailableData(msg StoreAvailableData) {
-	// TODO: handle store available data
+	err := av.availabilityStore.StoreAvailableData(msg.CandidateHash, msg.AvailableData)
+	if err != nil {
+		logger.Errorf("failed to load available data: %w", err)
+	}
```
Comment (kanishkatn) [inline dot/parachain/availability-store/availabilitystore.go:261]:
```suggestion
	if err != nil {
		logger.Errorf("load available data: %w", err)
	}
```

File: dot/parachain/availability-store/availabilitystore.go [hunk 4]
```diff
@@ -57,33 +172,92 @@ func (av *AvailabilityStoreSubsystem) processMessages() {
...
+		logger.Errorf("failed to load meta data: %w", err)
+		msg.Sender <- []ErasureChunk{}
+		return
+	}
+	chunks := []ErasureChunk{}
+	for i, v := range meta.ChunksStored {
+		if v {
+			chunk, err := av.availabilityStore.LoadChunk(msg.CandidateHash, uint32(i))
+			if err != nil {
+				logger.Errorf("failed to load chunk: %w", err)
```
Comment (kanishkatn) [inline dot/parachain/availability-store/availabilitystore.go:231]:
```suggestion
				logger.Errorf("load chunk: %w", err)
```

File: dot/parachain/availability-store/availabilitystore.go [hunk 5]
```diff
@@ -57,33 +172,92 @@ func (av *AvailabilityStoreSubsystem) processMessages() {
...
+		logger.Errorf("failed to load chunk: %w", err)
+	}
+	msg.Sender <- uint32(len(chunk.Chunk))
 }
 
 func (av *AvailabilityStoreSubsystem) handleQueryAllChunks(msg QueryAllChunks) {
-	// TODO: handle query all chunks
+	meta, err := av.availabilityStore.LoadMetaData(msg.CandidateHash)
+	if err != nil {
+		logger.Errorf("failed to load meta data: %w", err)
```
Comment (kanishkatn) [inline dot/parachain/availability-store/availabilitystore.go:222]:
```suggestion
		logger.Errorf("load metadata: %w", err)
```

File: dot/parachain/availability-store/availabilitystore.go [hunk 6]
```diff
@@ -57,33 +172,92 @@ func (av *AvailabilityStoreSubsystem) processMessages() {
...
+	for i, v := range meta.ChunksStored {
+		if v {
+			validatorIndex = uint32(i)
+			break
+		}
+	}
+
+	chunk, err := av.availabilityStore.LoadChunk(msg.CandidateHash, validatorIndex)
+	if err != nil {
+		logger.Errorf("failed to load chunk: %w", err)
```
Comment (kanishkatn) [inline dot/parachain/availability-store/availabilitystore.go:214]:
```suggestion
		logger.Errorf("load chunk: %w", err)
```

File: dot/parachain/availability-store/availabilitystore.go [hunk 7]
```diff
@@ -57,33 +172,92 @@ func (av *AvailabilityStoreSubsystem) processMessages() {
...
+	}
+	msg.Sender <- result
 }
 
 func (av *AvailabilityStoreSubsystem) handleQueryChunkSize(msg QueryChunkSize) {
-	// TODO: handle query chunk size
+	meta, err := av.availabilityStore.LoadMetaData(msg.CandidateHash)
+	if err != nil {
+		logger.Errorf("failed to load meta data: %w", err)
```
Comment (kanishkatn) [inline dot/parachain/availability-store/availabilitystore.go:202]:
```suggestion
		logger.Errorf("load metadata: %w", err)
```

**Review by kanishkatn — APPROVED** [review]

File: dot/node_integration_test.go (hunk 2 above)
Comment (edwardmack) [reply dot/node_integration_test.go:127]: Good question, I'm not sure, `phs` was defined before I modified this code.

File: dot/parachain/availability-store/messages.go
Comment (edwardmack) [reply dot/parachain/availability-store/messages.go:1]: Good idea, I've added comments

**Comment by kishansagathiya** [comment]
@edwardmack going through implementers guide and the code, I feel like we definitely need pruning. Feel free to create an issue for that.

**Comment by edwardmack** [comment]
> @edwardmack going through implementers guide and the code, I feel like we definitely need pruning. Feel free to create an issue for that.

I've created issue #3561

File: dot/parachain/availability-store/availabilitystore.go [hunk 8]
```diff
@@ -1,33 +1,159 @@
 // Copyright 2023 ChainSafe Systems (ON)
 // SPDX-License-Identifier: LGPL-3.0-only
```
Comment (kishansagathiya) [inline dot/parachain/availability-store/availabilitystore.go:3]: lets rename the file to availability_store.go

File: dot/parachain/availability-store/availabilitystore_test.go [hunk 9]
```diff
@@ -0,0 +1,226 @@
+package availabilitystore
+
```
Comment (kishansagathiya) [inline dot/parachain/availability-store/availabilitystore_test.go:2]: lets rename the file to availability_store_test.go

File: dot/parachain/availability-store/availabilitystore_test.go [hunk 10]
```diff
@@ -0,0 +1,226 @@
+package availabilitystore
```
Comment (kishansagathiya) [inline dot/parachain/availability-store/availabilitystore_test.go:1]: add licenses to this using `make license`

File: dot/parachain/availability-store/availabilitystore.go [hunk 11]
```diff
@@ -1,33 +1,159 @@
...
+}
+
+// NewAvailabilityStore creates a new instance of AvailabilityStore
+func NewAvailabilityStore(db database.Database) (*AvailabilityStore, error) {
+	return &AvailabilityStore{
+		availableTable: database.NewTable(db, avaliableDataPrefix),
+		chunkTable:     database.NewTable(db, chunkPrefix),
+		metaTable:      database.NewTable(db, metaPrefix),
+	}, nil
+}
```
Comment (kishansagathiya) [inline dot/parachain/availability-store/availabilitystore.go:53]:
```suggestion
func NewAvailabilityStore(db database.Database) *AvailabilityStore {
	return &AvailabilityStore{
		availableTable: database.NewTable(db, avaliableDataPrefix),
		chunkTable:     database.NewTable(db, chunkPrefix),
		metaTable:      database.NewTable(db, metaPrefix),
	}
}
```

File: dot/parachain/availability-store/availabilitystore.go [hunk 12]
```diff
@@ -1,33 +1,159 @@
...
+// loadAvailableData loads available data from the availability store
+func (as *AvailabilityStore) loadAvailableData(candidate common.Hash) (AvailableData, error) {
+	resultBytes, err := as.availableTable.Get(candidate[:])
+	if err != nil {
+		return AvailableData{}, err
+	}
+	result := AvailableData{}
+	err = json.Unmarshal(resultBytes, &result)
+	return result, err
+}
```
Comment (kishansagathiya) [inline dot/parachain/availability-store/availabilitystore.go:64]:
```suggestion
func (as *AvailabilityStore) loadAvailableData(candidate common.Hash) (*AvailableData, error) {
	resultBytes, err := as.availableTable.Get(candidate[:])
	if err != nil {
		return nil, fmt.Errorf("getting candidate %v from available table: %w", candidate, err)
	}
	result := AvailableData{}
	err = json.Unmarshal(resultBytes, &result)
	if err != nil {
		return nil, fmt.Errorf("unmarshalling available data: %w", err)
	}
	return &result, nil
}
```

- let's wrap errors
- lets return AvailableData as pointer

File: dot/parachain/availability-store/availabilitystore.go [hunk 13]
```diff
@@ -1,33 +1,159 @@
...
+// loadMetaData loads metadata from the availability store
+func (as *AvailabilityStore) loadMetaData(candidate common.Hash) (CandidateMeta, error) {
+	resultBytes, err := as.metaTable.Get(candidate[:])
+	if err != nil {
+		return CandidateMeta{}, err
+	}
+	result := CandidateMeta{}
+	err = json.Unmarshal(resultBytes, &result)
+	return result, err
+}
```
Comment (kishansagathiya) [inline dot/parachain/availability-store/availabilitystore.go:75]:
```suggestion
func (as *AvailabilityStore) loadMetaData(candidate common.Hash) (*CandidateMeta, error) {
	resultBytes, err := as.metaTable.Get(candidate[:])
	if err != nil {
		return nil, fmt.Errorf("getting candidate %v from available table: %w", candidate, err)
	}
	result := CandidateMeta{}
	err = json.Unmarshal(resultBytes, &result)
	if err != nil {
		return nil, fmt.Errorf("unmarshalling candidate meta: %w", err)
	}
	return &result, nil
}
```

File: dot/parachain/availability-store/availabilitystore.go [hunk 14]
```diff
@@ -1,33 +1,159 @@
...
+
+// storeMetaData stores metadata in the availability store
+func (as *AvailabilityStore) storeMetaData(candidate common.Hash, meta CandidateMeta) error {
+	dataBytes, err := json.Marshal(meta)
+	if err != nil {
+		return err
```
Comment (kishansagathiya) [inline dot/parachain/availability-store/availabilitystore.go:81]: error wrap

File: dot/parachain/availability-store/availabilitystore.go [hunk 15]
```diff
@@ -1,33 +1,159 @@
...
+	}
+	return as.metaTable.Put(candidate[:], dataBytes)
```
Comment (kishansagathiya) [inline dot/parachain/availability-store/availabilitystore.go:83]: error wrap

File: dot/parachain/availability-store/availabilitystore.go [hunk 16]
```diff
@@ -1,33 +1,159 @@
...
+// loadChunk loads a chunk from the availability store
+func (as *AvailabilityStore) loadChunk(candidate common.Hash, validatorIndex uint32) (ErasureChunk, error) {
+	resultBytes, err := as.chunkTable.Get(append(candidate[:], uint32ToBytes(validatorIndex)...))
+	if err != nil {
+		return ErasureChunk{}, err
+	}
+	result := ErasureChunk{}
+	err = json.Unmarshal(resultBytes, &result)
+	return result, err
+}
```
Comment (kishansagathiya) [inline dot/parachain/availability-store/availabilitystore.go:95]: - wrap errors 
- written pointer

Looks like this is happening at many places. I will avoid commenting everywhere.

File: dot/parachain/availability-store/availabilitystore.go [hunk 17]
```diff
@@ -1,33 +1,159 @@
...
+
+// storeChunk stores a chunk in the availability store
+func (as *AvailabilityStore) storeChunk(candidate common.Hash, chunk ErasureChunk) error {
+	meta, err := as.loadMetaData(candidate)
+	if err != nil {
+		if err.Error() == "pebble: not found" {
```
Comment (kishansagathiya) [inline dot/parachain/availability-store/availabilitystore.go:101]: use sentinel error for `pebble: not found`. see if you can use `errors.Is`, here.

File: dot/parachain/availability-store/availabilitystore.go [hunk 18]
```diff
@@ -1,33 +1,159 @@
...
+			meta = CandidateMeta{
+				ChunksStored: make([]bool, 16),
+			}
+		} else {
+			return err
+		}
+	}
+
+	if meta.ChunksStored[chunk.Index] {
+		return nil // already stored
```
Comment (kishansagathiya) [inline dot/parachain/availability-store/availabilitystore.go:113]: let's a debug log here  for index already stored, might be useful.

File: dot/parachain/availability-store/availabilitystore.go [hunk 19]
```diff
@@ -1,33 +1,159 @@
...
+			return err
+		}
+
+		meta.ChunksStored[chunk.Index] = true
+		err = as.storeMetaData(candidate, meta)
+		if err != nil {
+			return err
+		}
+	}
+	logger.Debugf("Stored chuck %i for %v", chunk.Index, candidate)
```
Comment (kishansagathiya) [inline dot/parachain/availability-store/availabilitystore.go:130]:
```suggestion
	logger.Debugf("stored chuck %d for %v", chunk.Index, candidate)
```

File: dot/parachain/availability-store/availabilitystore.go [hunk 20]
```diff
@@ -57,33 +183,93 @@ func (av *AvailabilityStoreSubsystem) processMessages() {
...
 }
 
 func (av *AvailabilityStoreSubsystem) handleQueryDataAvailability(msg QueryDataAvailability) {
-	// TODO: handle query data availability
+	_, err := av.availabilityStore.loadMetaData(msg.CandidateHash)
+	if err != nil {
+		msg.Sender <- false
+	} else {
+		msg.Sender <- true
+	}
```
Comment (kishansagathiya) [inline dot/parachain/availability-store/availabilitystore.go:199]:
```suggestion
msg.Sender <- err == nil
```

Comment (kishansagathiya) [reply dot/parachain/availability-store/availabilitystore.go:199]: We should also return this error.
We could ignore `pebble: not found` error, but should forward any other errors. And then log these errors in processMessages

File: dot/parachain/availability-store/availabilitystore.go [hunk 21]
```diff
@@ -57,33 +183,93 @@ func (av *AvailabilityStoreSubsystem) processMessages() {
...
+			break
+		}
+	}
+
+	chunk, err := av.availabilityStore.loadChunk(msg.CandidateHash, validatorIndex)
+	if err != nil {
+		logger.Errorf("load chunk: %w", err)
+	}
+	msg.Sender <- uint32(len(chunk.Chunk))
 }
```
Comment (kishansagathiya) [inline dot/parachain/availability-store/availabilitystore.go:228]: No need to log errors here. Let's  return errors in all `handle` functions and log them in processMessages. 
That way we only log them once in the end.

will avoid adding this comment at other such instances.

File: dot/parachain/availability-store/availabilitystore.go [hunk 22]
```diff
@@ -57,33 +183,93 @@ func (av *AvailabilityStoreSubsystem) processMessages() {
...
+	chunks := []ErasureChunk{}
+	for i, v := range meta.ChunksStored {
+		if v {
+			chunk, err := av.availabilityStore.loadChunk(msg.CandidateHash, uint32(i))
+			if err != nil {
+				logger.Errorf("load chunk: %w", err)
+			}
+			chunks = append(chunks, chunk)
+		} else {
+			logger.Warnf("chunk %i not stored for %v", i, msg.CandidateHash)
```
Comment (kishansagathiya) [inline dot/parachain/availability-store/availabilitystore.go:246]:
```suggestion
			logger.Warnf("chunk %d not stored for %v", i, msg.CandidateHash)
```

File: dot/parachain/availability-store/availabilitystore_test.go [hunk 23]
```diff
@@ -0,0 +1,226 @@
...
+
+	msg2 := QueryChunkAvailability{
+		CandidateHash:  common.Hash{0x01},
+		ValidatorIndex: 2,
+		Sender:         msgSenderChan,
+	}
+	go asSub.handleQueryChunkAvailability(msg2)
+	msgSenderChanResult2 := <-msg2.Sender
+	require.Equal(t, false, msgSenderChanResult2)
+}
```
Comment (kishansagathiya) [inline dot/parachain/availability-store/availabilitystore_test.go:186]: let's use test cases here

one example here https://github.com/ChainSafe/gossamer/blob/63402322506fcda7e23120c70e4ac8db0f6e9fd6/dot/core/service_integration_test.go#L106

File: dot/node_integration_test.go (hunk 2 above)
Comment (kanishkatn) [reply dot/node_integration_test.go:127]: parachain host service?

File: dot/parachain/availability-store/availabilitystore_test.go (hunk 23 above)
Comment (edwardmack) [reply dot/parachain/availability-store/availabilitystore_test.go]: Updated tests.

**Pull request merged**